### PR TITLE
ext_authz: optionally have HTTP response for GRPC calls reflect GRPC status

### PR DESCRIPTION
**Pull request by dclendenan**
Commit Message:
New ext_authz config item `http_status_for_grpc_deny`, if true, will tell Envoy to have the HTTP
status for GRPC calls reflect the GRPC status.  If false or unset, no change.

Possible solution for https://github.com/envoyproxy/envoy/issues/11079

Additional Description:
Since the path between the ext-authz layer and sendLocalReply() is
somewhat indirect, a temporary header is used as a configuration
conduit.

Thanks @ramaraochavali for guidance.

Risk Level: Low

Testing:
New Unit and Integration tests added to cover positive and negative cases.

Manual testing was done by modifying the ext_authz example configuration and using Curl to send requests.

Run modified example code with locally-build envoy:
 cp -f ../../bazel-bin/source/exe/envoy-static . && docker-compose build --parallel && docker-compose up

Curl command (negative, use GET for allowed request):
curl -v -X POST -H "Authorization: Bearer token1"  -H "Content-Type:application/grpc" localhost:8000/service

Locally modified files for testing:
```
--- ../ext_authz/Dockerfile-frontenvoy  2021-04-14 15:13:42.026126152 -0700
+++ Dockerfile-frontenvoy       2021-04-13 16:36:19.138377072 -0700
@@ -1,9 +1,17 @@
 FROM envoyproxy/envoy-dev:latest

 RUN apt-get update && apt-get -q install -y \
     curl
 COPY ./config /etc/envoy-config
 COPY ./run_envoy.sh /run_envoy.sh
+COPY ./envoy-static /usr/local/bin/envoy
 RUN chmod go+r -R /etc/envoy-config \
     && chmod go+rx /run_envoy.sh /etc/envoy-config /etc/envoy-config/*
 CMD ["/bin/sh", "/run_envoy.sh"]


--- ../ext_authz/config/opa-service/v3.yaml     2021-04-14 15:13:42.026126152 -0700
+++ config/opa-service/v3.yaml  2021-04-13 16:39:49.173523268 -0700
@@ -24,20 +24,22 @@ static_resources:
                   cluster: upstream-service
           http_filters:
           - name: envoy.filters.http.ext_authz
             typed_config:
               "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
               grpc_service:
                 envoy_grpc:
                   cluster_name: ext_authz-opa-service
                 timeout: 0.250s
               transport_api_version: V3
+              http_status_for_grpc_deny: true
+              #http_status_for_grpc_deny: false
           - name: envoy.filters.http.router
             typed_config: {}
```

Docs Changes:  Pending

Release Notes:  Pending

Platform Specific Features:  N/A

### CODE REVIEW

**Review by rojkov — COMMENTED** [review]
Thank you for working on this! I added a few questions to clarify.

File: api/envoy/config/filter/network/ext_authz/v2/ext_authz.proto
Comment (htuch) [inline api/envoy/config/filter/network/ext_authz/v2/ext_authz.proto:53]: No changes to v2 are allowed, these APIs are frozen (Envoy will drop support completely for v2 after the 1.18.0 release going out today).

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:130]: This field doesn't reflect the comments; are we always reflecting the grpc-status or only on deny?

**Comment by htuch** [comment]
@esmet can you take a pass on this? Thanks.

File: api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto
Comment (esmet) [inline api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto:117]: Does this mean Envoy will send back the canonical HTTP status code for the observed gRPC status? i.e. by using `httpToGrpcStatus` from source/common/grpc/status.h?

File: source/common/http/utility.cc
Comment (esmet) [inline source/common/http/utility.cc:521]: I feel like this should be using httpToGrpcStatus on the observed grpc-status from the authorization server. Either that, or I misunderstand the original intention.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (dclendenan) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: The ext_authz configuration and state aren't all available in the sendLocalReply() where this is being used.   I agree that the header indirection is a tad convoluted, I'm open to suggestions for an alternate mechanism that accomplishes what is called out in https://github.com/envoyproxy/envoy/issues/11079 - barring that there are definitely some more docs required for this change.

Comment (esmet) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: Right, my sense is that we should be able to do this without using a header value as control flow (and I would prefer to not use a header if possible).

For example, `statusOnError` and the other calls to sendLocalReply show precedent for setting the http status code to return to the downstream client. I bet we can just pass it below to decoder_callbacks_->sendLocalReply near L306.

File: source/extensions/filters/http/ext_authz/ext_authz.h
Comment (esmet) [inline source/extensions/filters/http/ext_authz/ext_authz.h:81]: If this is indeed the canonical http status for grpc deny cases, I would add the word canonical in here and use the word 'on' instead of 'for' (precedent is `status_on_error`, IMO)

`canonicalHttpStatusOnGrpcDeny`

File: source/common/http/utility.cc
Comment (dclendenan) [reply source/common/http/utility.cc:521]: I missed that function when I made the first pass at this, but as @rojkov noted above, it's simpler to just only reset as ...OK if this option isn't set. Next push will have that.

Comment (esmet) [reply source/common/http/utility.cc:521]: Has `response_headers` already been mutated at this point to contain the canonical http response code for the observed upstream grpc-status?

Comment (esmet) [reply source/common/http/utility.cc:521]: Oh I see that this is a utility function. In that case then yes, the response_headers_ really do need to be updated by this point.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (dclendenan) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:130]: It's currently on-deny, the comments reflect a prior state in the PR before input from @ramaraochavali.  Will align them based on feedback here, thanks for calling that out.

Comment (ramaraochavali) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:130]: We want only on deny as described in the issue - IMO making it for all status is too broad and needed.

**Comment by esmet** [comment]
After reading the description and looking at this a lot closer, I think I now better understand the use case, which is to send back the canonical HTTP status of the authorization server's grpc-status to a _gRPC client_. I was originally confused about the last part.

I feel like this reduces to having the ability to tell sendLocalReply which HTTP status to use for gRPC clients. Currently it sends back OK as long as `local_reply_data_.is_grpc_` is set (meaning the client is gRPC). Maybe we need to extend local_reply_data_ to encode whether the caller wants the canonical HTTP status instead of just OK. Let me look a little deeper into that.

**Comment by dclendenan** [comment]
@esmet the specific case being addressed is replies to gRPC clients which are rejected due to OPA rules, and users are confused by the 200-OK status.  A more general change might end up better, and require a less complex approach - thanks for taking the time to explore further.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (ramaraochavali) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: +1 for not using header if possible. But `statusOnError` is only honoured for http requests I guess because for gRPC it is overwritten https://github.com/envoyproxy/envoy/blob/main/source/common/http/utility.cc#L520 - So I think it is not possible without changing signature to let sendLocalReply in filter manager and utility know that it should not be overwritten,

Comment (esmet) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: Yes after looking harder, I agree that sendLocalReply would need to be expanded to include the desired HTTP code for gRPC clients.

Comment (dclendenan) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: So the consensus is that we should take the more explicit path of changing call signatures up the chain of `sendLocalReply()` variants to allow explicitly passing the HTTP response code?  
I will revert the changes to the V2 config and get started on replacing the header-magic with the more explicit path for V3+.

Comment (esmet) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: I think sendLocalReply should support sending back the translated grpcToHttpStatus instead of just 200 OK. We can then configure the components that use sendLocalReply (ext_authz in this case, but in theory ratelimit as well, or even on the HttpConnectionManager which would apply the behavior to all downstreams for that HCM).

There are some signatures to modify going down, but eventually it becomes LocalReplyData which I think is a bit less painful to deal with.

Comment (dclendenan) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: There are some mocks to modify if we change the signatures for the various `sendLocalReply()` functions, but it's relatively straightforward.  However, if we have `Utility::sendLocalReply()` call `grpcToHttpStatus()` there appear to be some dependency-cycle issues to get access to that function in grpc/common - any issue with just passing the HTTP code itself up the chain rather than a flag to tell `Utility::sendLocalReply()`  to figure out what code to use?

My first cut, which should be ready to push tomorrow, has the code passed along as an optional parameter until the `LocalReplyData` object is available, and then bundles it within that as a new variable - that seem sane to you?

Comment (esmet) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: I would prefer to not add another http code parameter when one already exists.  I think a fine first cut would be to pass down a new boolean that says whether to use the http status for gRPC clients (no need for grpcToHttpStatus in this cut). This is similar to what you already have, but instead of the call to httpCodeForGrpcDeny it would just be checking that piped down parameter.

**Comment by dclendenan** [comment]
Did the linter rules change?  That function name predates this PR, but seems to make it unhappy now, so I changed it to something that seemed acceptable.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (dclendenan) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:305]: Latest push adds that boolean, it's a new parameter for the steps before `LocalReplyData` is available.  Unit tests and manual via my modified ext_auth example folder are all passing.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (dclendenan) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto]: comment updated, and formatted as per comment above.

File: source/common/http/utility.h
Comment (dclendenan) [reply source/common/http/utility.h]: this function was factored out along with the header indirection.

Comment (dclendenan) [reply source/common/http/utility.h]: This function has been removed.  Thanks for the review, I think I've updated everything you'd called out.

File: include/envoy/http/filter.h
Comment (esmet) [inline include/envoy/http/filter.h:835]: I think this should say 'bool' and that it is a flag, noting that the code likely comes from the response headers.

File: source/common/http/utility.h
Comment (esmet) [inline source/common/http/utility.h:317]: Can we clarify whether the the called service's response means 'comes from the response headers object'?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (esmet) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:307]: Not totally sure we need this logging

File: api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto
Comment (esmet) [inline api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto]: I think  we should update this comment to say that when this is set to true, the downstream HTTP response code will match the HTTP response code observed from the authorization server instead of 200 OK. We should also clarify whether grpc-status (and other grpc headers) are still maintained, since I think the outcome of using this flag is a sort of mixed-mode gRPC/HTTP reply (some semantics of grpc, some of HTTP 1?)

Comment (dclendenan) [reply api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto]: I did update this somewhat in my last push, based on @rojkov's comment but can clarify it further w.r.t. gRPC vs HTTP.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (dclendenan) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:307]: this is a hangover from the header-conduit approach that I missed, will remove.

File: source/extensions/filters/network/ext_authz/ext_authz.h
Comment (dclendenan) [reply source/extensions/filters/network/ext_authz/ext_authz.h]: this has evolved somewhat, but it's called in `ext_authz.cc:onComplete()`

File: api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto
Comment (dclendenan) [reply api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto]: I've updated the comments, as suggested.  What is the process for updating the docs?  I didn't see a reference in the contributing reference, but may have just missed it.

Thanks for the extensive and speedy reviews!

Comment (rojkov) [reply api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto]: The API docs get updated automatically upon changes in proto files. For example the doc page generated from this proto is https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#envoy-v3-api-msg-extensions-filters-http-ext-authz-v3-extauthz.

To check how a doc page would look like after your changes you can generate it locally with `./docs/build.sh`.

Comment (dclendenan) [reply api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto]: Thanks @rojkov - I've updated the proto notes based on the generation results.

**Comment by dclendenan** [comment]
Note that after the merge I am seeing a DCO error from a change that has merged into `main`.  Per the developer docs I haven't rebased it away, but I can if it causes CI issues.

I needed to get caught up with `main` to fix a test issue that I saw in CI but not in my local before the merge.

```
ERROR: The following commits do not have DCO signoff:
  ad4919dd7 upstream adding QUIC-to-TCP failover (#15894)
```

File: api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto
Comment (dclendenan) [reply api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto]: `links -dump  generated/docs/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto.html |grep -A9 retain_http_status_on_grpc_deny` 

-->

```
   retain_http_status_on_grpc_deny

           (bool) Specifies whether HTTP status codes for gRPC calls are
           passed intact from the called service back to the caller on gRPC
           DENIED.

           When this field is true, Envoy’s HTTP response to gRPC calls which
           receive a DENIED response will retain the status code from the
           called service’s HTTP headers. The gRPC headers are not affected.
```

**Comment by dclendenan** [comment]
I see these doc-build warnings in CI and my local build since I merged in the main branch, I'm pretty sure they're not caused by my change.   I'm happy to add a fix for the warnings but would need some guidance on the approach to take.

File: api/BUILD
Comment (rojkov) [inline api/BUILD:147]: I guess these changes are not relevant to this PR and explain the doc generation failures.

Comment (dclendenan) [reply api/BUILD]: Thanks @rojkov - I've verified that the api/BUILD change was the culprit.  It was triggered (but not committed) by the git-push automation, `./tools/proto_format/proto_format.sh fix` also causes it.   It's reverted for now in my feature branch, to get CI passing for this PR.

**Comment by dclendenan** [comment]
@esmet  I've resolved the conflicting changes from new merges and all tests are passing.  Do you have any other questions or comments on this PR?

Thanks all for detailed reviews!

**Comment by dclendenan** [comment]
@htuch  @PiotrSikora can you please have a look?  

FYI I've  re-kicked CI since the last run failed what appears to be a flapping test on windows, all other platforms passed.

File: source/common/http/utility.h
Comment (htuch) [inline source/common/http/utility.h:333]: I think these code comments and even the nature of this boolean are a bit confusing. What is the "called service"? This makes sense in ext_authz, when we make a RPC sidecall. What does this mean for other filters etc.

The main concern here is that we are leaking ext_authz details into Envoy core.

Comment (dclendenan) [reply source/common/http/utility.h:333]: Would you prefer to have it say "whether to retain HTTP status in response headers for gRPC calls" and remove the ambiguous/wordy "called service"?  

It's sending an ext_authz decision back into Envoy core, by way of the boolean.  Other filters might leverage this decision-passing mechanism differently, but I agree that it's constrained by having been designed for this ext_authz use case.

File: include/envoy/http/filter.h
Comment (htuch) [inline include/envoy/http/filter.h:437]: Why isn't this an options struct as elsehwere? A little uncomfortable proliferating more args here.

File: source/common/http/utility.h
Comment (htuch) [reply source/common/http/utility.h:333]: I think it might be clearer to describe it in terms of other features of this options struct. Namely `if is_grpc_ and retain_http_status_for_grpc_ is set, then the HTTP status code will be response_code_ instead of the gRPC standard 200`

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:130]: Which gRPC headers? What if the client call was gRPC? It's helpful to be very specific here.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:126]: `from the called ext_authz service`

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:125]: Can you elaborate on when this is used? Usually an ext_authz gRPC service will return 200, so what scenario does this matter? I'm asking for the reader, so updating the comments here makes sense; I personally am familiar with when gRPC can return non-200 (e.g. early failure), but less clear to docs reader.

**Comment by alyssawilk** [comment]
Dio (owner) is out.  Greg can you do passes until Harvey gets back?

File: source/common/http/utility.cc
Comment (ggreenway) [inline source/common/http/utility.cc:532]: I don't think this is correct. According to the grpc spec, grpc errors are communicated through the grpc code, and http-200 should be present in the response even on failure.

File: source/extensions/filters/network/ext_authz/ext_authz.h
Comment (ggreenway) [inline source/extensions/filters/network/ext_authz/ext_authz.h:57]: This is set here but I don't think it is ever used. What would it even mean for the network filter? My guess is that this only applies to http ext_auth.

**Review by ggreenway — CHANGES_REQUESTED** [review]
I agree with @htuch (https://github.com/envoyproxy/envoy/pull/15979#discussion_r641834916) that an options struct might be appropriate for this type of change.

/wait

File: source/common/http/utility.cc
Comment (dclendenan) [reply source/common/http/utility.cc:532]: Are you saying you don't think the comment is correct, or the change to make the HTTP code match the DENIED response?

Comment (ggreenway) [reply source/common/http/utility.cc:532]: I'm saying that according to the gRPC spec, both success and failure of the gRPC call should have http-200 as the response code, so the comment `Default 'OK' reflects successful gRPC call` isn't correct, because `OK` doesn't mean it succeeded.

Comment (dclendenan) [reply source/common/http/utility.cc:532]: Thanks for clarifying, I'll correct the text.

File: include/envoy/http/filter.h
Comment (htuch) [reply include/envoy/http/filter.h:438]: Friendly ping

File: source/extensions/filters/network/ext_authz/ext_authz.h
Comment (htuch) [reply source/extensions/filters/network/ext_authz/ext_authz.h:57]: Friendly ping

**Review by htuch — COMMENTED** [review]
Going back to https://github.com/envoyproxy/envoy/issues/11079, I'm still not sure why we need this. Can you remind me what real world context this comes up in?

**Comment by dclendenan** [comment]
I'm back from vacation, I think at this point we are back to some design discussions in the issue (https://github.com/envoyproxy/envoy/issues/11079#issuecomment-870742887) so it probably makes sense to hold off on further changes in this PR until we've come to consensus there as to the approach.  

@markdroth suggested an alternative, can others please comment in the issue on whether that seems reasonable/preferred?

/wait

**Comment by htuch** [comment]
I like the idea of building the mechanism for the server to specify gRPC/HTTP status codes and leaving policy to the server.

**Comment by dclendenan** [comment]
withdrawing this, we will look for a less invasive way to resolve the issue in a new PR.

**Pull request closed without merging**
